packages/service-provider-core/src/connection-string.ts

```diff
@@ -0,0 +1,112 @@
+    } catch (err) {
+      throw new MongoshInvalidInputError(err.message, CommonErrors.InvalidArgument);
+    }
+
+    // characters not permitted in username nor password Set([':', '/', '?', '#', '[', ']', '@'])
+    const illegalCharacters = new RegExp(String.raw`[:/?#\[\]@]`, 'gi');
+    if (username?.match(illegalCharacters)) {
+      throw new MongoshInvalidInputError(`Username contains unescaped characters ${username}`, CommonErrors.InvalidArgument);
+    }
+    if (!username || !password) {
+      const uriWithoutProtocol = uri.replace(`${protocol}://`, '');
+      if (uriWithoutProtocol.startsWith('@') || uriWithoutProtocol.startsWith(':')) {
+        throw new MongoshInvalidInputError('URI contained empty userinfo section', CommonErrors.InvalidArgument);
+      }
+    }
+
+    if (password?.match(illegalCharacters)) {
+      throw new MongoshInvalidInputError('Password contains unescaped characters', CommonErrors.InvalidArgument);
+    }
+
+    let authString = '';
+    if (typeof username === 'string') authString += username;
+    if (typeof password === 'string') authString += `:${password}`;
+
+    super(`${protocol.toLowerCase()}://${authString}@${DUMMY_HOSTNAME}${rest}`);
+    this._hosts = hosts.split(',');
+
+    if (this.isSRV && this.hosts.length !== 1) {
+      throw new MongoshInvalidInputError('mongodb+srv URI cannot have multiple service names', CommonErrors.InvalidArgument);
+    }
+    if (this.isSRV && this.hosts.some(host => host.includes(':'))) {
+      throw new MongoshInvalidInputError('mongodb+srv URI cannot have port number', CommonErrors.InvalidArgument);
+    }
+    if (!this.pathname) {
+      this.pathname = '/';
+    }
+  }
+
+  // The getters here should throw, but that would break .toString() because of
+  // https://github.com/nodejs/node/issues/36887. Using 'never' as the type
+  // should be enough to stop anybody from using them in TypeScript, though.
+  get host(): never { return DUMMY_HOSTNAME as never; }
+  set host(_ignored: never) { throw new MongoshInternalError('No single host for connection string'); }
+  get hostname(): never { return DUMMY_HOSTNAME as never; }
+  set hostname(_ignored: never) { throw new MongoshInternalError('No single host for connection string'); }
+  get port(): never { return '' as never; }
+  set port(_ignored: never) { throw new MongoshInternalError('No single host for connection string'); }
+  get href(): string { return this.toString(); }
+  set href(_ignored: string) { throw new MongoshInternalError('No single host for connection string'); }
+
+  get isSRV(): boolean {
+    return this.protocol.includes('srv');
+  }
+
+  get hosts(): string[] {
+    return this._hosts;
+  }
+
+  set hosts(list: string[]) {
+    this._hosts = list;
+  }
+
+  toString(): string {
+    return super.toString().replace(DUMMY_HOSTNAME, this.hosts.join(','));
+  }
+
+  clone(): ConnectionString {
+    return new ConnectionString(this.toString());
+  }
+}
```

Comment:
Talked with @mcasimir and, if we like this, we can also split this out into a separate package that compass and/or the driver itself could then use

Reply:
@Anemy you may want to take a look if this is something viable for compass